docs/adrs/0549-composite-run-steps.md

```diff
@@ -0,0 +1,273 @@
+
+#### Exposing Parent's If Condition to Children Via a Variable
+It would be nice to have a way to access information from a parent's if condition. We could have a parent variable that is contained in the context similar to other context variables `github`, `strategy`, etc.:
+
+Example `user/composite/action.yml`
+steps:
+  - run: echo "preparing the slack bot..."  # <--- This will run, as nothing has failed within the composite yet
+  - run: slack.post("All builds passing, ready for a deploy")  # <-- this will not run, as the parent fails
+    if: ${{ parent.success() }}
+  - run: slack.post("A failure has happened, fix things now", alert=true)  # <--- This will run, as the parent fails
+    if: ${{ parent.failure() }}
+
+Example `workflow.yml`
+steps:
+  - run: exit 1
+  - uses: user/composite@v1
+    if: always()
+
+    
+### Timeout-minutes
+Example `user/composite/action.yml`:
+```
+using: 'composite' 
+steps: 
+  - id: foo1
+    run: echo test 1
+    timeout-minutes: 10
+  - id: foo2
+    run: echo test 2
+  - id: foo3
+    run: echo test 3
+    timeout-minutes: 10
+```
+
+Example `workflow.yml`:
+```
+steps: 
+  - id: bar
+    uses: user/test@v1
+    timeout-minutes: 50
+```
+**TODO: This timeout-minutes condition implementation is up to discussion.** 
+
+A composite action in its entirety is a job. You can set both timeout-minutes for the whole composite action or its steps as long as the the sum of the `timeout-minutes` for each composite action step that has the attribute `timeout-minutes` is less than or equals to `timeout-minutes` for the composite action. There is no default timeout-minutes for each composite action step. 
+
+If the time taken for any of the steps in combination or individually exceed the whole composite action `timeout-minutes` attribute, the whole job will fail. If an individual step exceeds its own `timeout-minutes` attribute but the total time that has been used including this step is below the overall composite action `timeout-minutes`, the individual step will fail but the rest of the steps will run. 
+
+For reference, in the example above, if the composite step `foo1` takes 11 minutes to run, that step will fail but the rest of the steps, `foo1` and `foo2`, will proceed as long as their total runtime with the previous failed `foo1` action is less than the composite action's `timeout-minutes` (50 minutes). If the composite step `foo2` takes 51 minutes to run, it will cause the whole composite action job to fail. I
+
+The rationale behind this is that users can configure their steps with the `if` condition to conditionally set how steps rely on each other. Due to the additional capabilities that are offered with combining `timeout-minutes` and/or `if`, we wanted the `timeout-minutes` condition to be as dumb as possible and not effect other steps. 
+
+[Usage limits still apply](https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions?query=if%28%29#usage-limits)
+
+
+### Continue-on-error
```

Comment:
There are a number of fields that can be populated for a step. Some of these are shared between `uses` and `runs`, and some are specific to one or the other (`shell` for `runs`, `with` for `uses`). Are we going to support overriding all of these by passing them into the composite action? `shell` and `working-directory`, can also be found in the job/workflow defaults defaults

It might be good to have a section on priority. As I understand it the order of precedence is:
- The value defined for an individual step in the action.yaml
- The value defined for the composite action in the workflow.yaml
- The value defined for the job defaults in the workflow.yaml
- The value defined for the workflow level defaults in the workflow.yaml

For example, can I do this?

Example `action.yml`
```
using: 'composite' 
steps: 
  - id: foo1
    run: env #only runs in powershell
    
```

Example `workflow.yml`:
```
steps: 
  - id: foo
    uses: user/composite@v1
    shell: powershell
```

We can talk about if, continue-on-error, timeout-minutes, and the like in that section as examples, rather then providing separate sections for them. Some of these values may also apply to the actual composite action as a whole as well (like timeout-minutes, if), so we can mention those also apply to the composite action rather then being just pass thru.

Reply:
We spoke about this offline.

we aren't planning on passing through any of the values that are passed into the composite action directly to each step inside the composite action. As a design principle, we are hoping workflow authors do not need to understand the implementation of the composite action, but can use it as a black box to accomplish what they want. If there needs to be input from the workflow author inside a composite action, they can use the `inputs` feature